src/js/components/Breadcrumb.js

```diff
@@ -0,0 +1,295 @@
+
+    return namedRoutes;
+  }
+
+  shouldCollapse(availableWidth, expandedWidth) {
+    // Smooth collapse action to prevent flickering between states
+    if (this.state.collapsed) {
+      return expandedWidth >= availableWidth - COLLAPSE_BUFFER;
+    }
+
+    return expandedWidth >= availableWidth + COLLAPSE_BUFFER;
+  }
+  /**
+   * Builds a crumb or multiple crumbs from the label and matching route
+   * @param  {String} label Url path part
+   * @param  {String} route Matching route name
+   * @param {Object} routeParams Router routeParams Object
+   * @param {Bool} labelIsParam Boolean indicating if the label represents a
+   * route parameter.
+   * @return {Array} Array of crumbs containing the label, route and route
+   * params. The returned label is the text shown for the crumb. The route is
+   * the route name which will be set on the Link-to attribute. The params will
+   * also be set on the Link tag. Each returned crumb will be spliced into the
+   * resulting breadcrumb.
+   */
+  buildCrumb(label, route, routeParams, labelIsParam) {
+    // Split label on whitespace and -
+    let splitLabelRegex = /[\s-]+/;
+    let joinLabelCharacter = ' ';
+    let params = {};
+
+    if (labelIsParam) {
+      // Extract param value as label e.g. (:userID -> Foo)
+      // and add param to params.
+      //
+      // Remove colon first
+      let paramName = label.slice(1);
+      // Decode value
+      label = decodeURIComponent(routeParams[paramName]);
+      params[paramName] = label;
+    } else {
+      // Split Label on space and - then capitalize each word and join.
+      // /foo-bar/ --> Foo Bar
+      label = label.split(splitLabelRegex).map(function (word) {
+        return StringUtil.capitalize(word);
+      }).join(joinLabelCharacter);
+    }
+    // Can return multiple route objects to get rendered in crumbs.
+    // Useful for splitting a route parameter like a file path.
+    return [{
+      label,
+      route,
+      params
+    }];
+  }
+
+  renderCrumbsFromRoute() {
+    let {router} = this.context;
+    let {buildCrumb, shift} = this.props;
+
+    if (buildCrumb == null) {
+      buildCrumb = this.buildCrumb;
+    }
+
+    let crumbKey = 1;
+    let namedRoutes = this.getCurrentNamedRoutes();
+    let params = router.getCurrentParams();
+    // Remove n items from beginning
+    let paths = Object.keys(namedRoutes).slice(shift);
+
+    return paths.reduce((crumbs, path, pathIndex) => {
+      let name = namedRoutes[path];
+      let pathParts = path.split('/');
+      // Decode Label
+      let label = decodeURIComponent(pathParts[pathParts.length - 1]);
+      let labelIsParam = label.charAt(0) === ':' && label.slice(1) in params;
+
+      let crumbObjects = buildCrumb(
+        label,
+        name,
+        params,
+        labelIsParam
+      );
+
+      crumbObjects.forEach((crumbParams, crumbIndex) => {
+        let {label, route, params} = crumbParams;
+
+        crumbs.push(
+          <li key={crumbKey++}>
```

Comment:
Could we move the view-parts to their own functions? e.g. `getLink` and `getArrow`, that should trim this function a little bit down in size and remove the `// Push divider ( > )` comment, since that should then be obvious